### DISCUSSION TITLE
Is there a way to add a hook or replace the ui command with a pipeline?

### DISCUSSION THREAD

**Discussion by rafaeljusto** [devspace-sh/devspace#2655] [Q&A] [answered]
Hi,

When running `devspace ui` I would like to add some security checks regarding the current K8s context, just to make sure the developer has the correct kubeconfig file. I couldn't find any [hook](https://www.devspace.sh/docs/5.x/configuration/hooks/basics#lifecycle-events) to intercept the `ui` call or a [default pipeline](https://www.devspace.sh/docs/configuration/pipelines/#default-pipelines) to replace it with extra commands. Is there an alternative here?

Thank you.

**Comment by lizardruss** [comment]
You could probably add a [custom command](https://www.devspace.sh/docs/configuration/commands/) to set the k8s context before running `devspace ui`.

**Reply by rafaeljusto** [reply]
Thank you for the quick answer! 🙂 Custom command looks like an interesting feature. On this case I would need to create a convention with the developers to run this command before executing `devspace ui`? That should be fine if there're no alternatives, I was just looking for something that would look transparent for the devs without adding more steps to the flow.

**Comment by jackwilsdon** [comment, marked as answer]
There appears to be a `ui` hook you could use: https://github.com/devspace-sh/devspace/blob/49e5f6bf1fd2fd127ddb0bce871413eec962ab09/cmd/ui.go#L181

```yaml
hooks:
  - name: pre-ui
    command: "echo before ui"
    events: ["ui"]
```

Returning a non-zero status code from the command causes the UI to not start:

```yaml
hooks:
  - name: pre-ui
    command: "echo before ui; exit 1"
    events: ["ui"]
```

```text
$ devspace ui        
info Using namespace 'default'
info Using kube context 'docker-desktop'
info Execute hook 'pre-ui' at ui
before ui
fatal in hook 'pre-ui': exit status 1
```

**Reply by rafaeljusto** [reply]
That's amazing! Looks like it was added [here](https://github.com/devspace-sh/devspace/pull/1215) ([commit](https://github.com/devspace-sh/devspace/commit/3dc21666b0a06ae08fa236494bb5999dfdfdaf35)) and never documented. Thank you! 🙂
